### ISSUE TITLE
BalancedConsumer raises an exception with not much details

### ISSUE THREAD

**Issue by tigran-a** [Parsely/pykafka#369]
Hi,

I've a consumption loop 

```
#!/usr/bin/env python

import pykafka
from consumeconfig import *
import time

import logging
log= logging.getLogger(__file__)

console_handler = logging.StreamHandler()
console_handler.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
console_handler.setFormatter(formatter)
logging.getLogger("pykafka").addHandler(console_handler)

log.addHandler(console_handler)
GROUP = "GroupB"
client = pykafka.KafkaClient(hosts=KAFKA_HOSTS)
topic = client.topics[TOPIC]

consumer  = topic.get_balanced_consumer(consumer_group=GROUP, zookeeper_connect=ZOOKEEPER_HOSTS, auto_commit_enable=True)

need_reconn  = False
cntr = 0
while True:
        try:
                m = consumer.consume()
                if m is not None:
                        print ("%d | %s."%(m.offset,m.value))
                        print ("\t\t\t" + str( consumer.held_offsets) )
                        cntr+=1
                if cntr % 100 == 0:
                        consumer.commit_offsets()
        except Exception, e:
                log.error(10*"*"+"|" + str(e) + "|"+10*"*")
```

which works fine... for some time. After an exception is raised and I can't see any details of it. 
Here is a log: 
```
No handlers could be found for logger "kazoo.client"
2015-11-24 13:30:15,931 - pykafka.balancedconsumer - WARNING - Internal partition registry doesn't match ZooKeeper!
2015-11-24 19:01:30,732 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,733 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,734 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,735 - ./consume-many.py - ERROR - **********||**********
2015-11-24 19:01:30,735 - ./consume-many.py - ERROR - **********||**********
...
```

Any hint what's going on (or what I'm doing wrong) ?

Thanks

**Comment by yungchin** [comment]
Thanks for the report @tigran-a - could you try this again with `repr(e)` instead of `str(e)` and `log.exception()` instead of `log.error()`? That should print a few more details. Plus if you set something like `logging.basicConfig(level=logging.DEBUG)` that should take care of the root logger, so that you can also see the kazoo errors.

**Comment by tigran-a** [comment]
Hi, I'm doing some tests in parallel, so it's another exception that's raised (right after that "empty" one):

```
2015-11-25 03:58:54,113 - pykafka.balancedconsumer - ERROR - Exception encountered in worker thread:
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 241, in checker 
    if not self._check_held_partitions():                                                
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 619, in _check_held_partitions 
    zk_partitions = self._get_held_partitions()                                          
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 600, in _get_held_partitions 
    all_partitions = self._zookeeper.get_children(self._topic_path)                      
  File "/usr/local/lib/python2.7/dist-packages/kazoo/client.py", line 1079, in get_children
    return self.get_children_async(path, watch, include_data).get()                           
  File "/usr/local/lib/python2.7/dist-packages/kazoo/handlers/utils.py", line 78, in get          
    raise self._exception   
```

I've had two balanced consumers running in parallel. One has failed with that error (the code does not exit on it, but tries to consume again, as nothing would happen, I just catch and print the exception). Another continued to work, but no messages from a partition that was used by the failed one. When I interrupted the failed, the other one took over those partitions.

Is it a kazoo problem? any workarounds?

**Comment by tigran-a** [comment]
What about wrapping all calls to _zookeeper in try/except and reconnect in case of problem;
especially that one: https://github.com/Parsely/pykafka/blob/master/pykafka/balancedconsumer.py#L642
It looks to me like a kazoo problem: https://github.com/python-zk/kazoo/issues/357, 
but maybe there are ways to overcome it?

**Comment by tigran-a** [comment]
Some more logs with repr:

```
2015-11-25 03:58:40,805 - kazoo.client - WARNING - Connection dropped: outstanding heartbeat ping not received
2015-11-25 03:58:40,805 - kazoo.client - WARNING - Transition to CONNECTING
2015-11-25 03:58:41,932 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 03:58:41,933 - kazoo.client - WARNING - Transition to CONNECTING
2015-11-25 03:58:47,915 - kazoo.client - WARNING - Connection dropped: outstanding heartbeat ping not received
2015-11-25 03:58:47,915 - kazoo.client - WARNING - Transition to CONNECTING
2015-11-25 10:01:16,002 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 10:01:16,002 - kazoo.client - WARNING - Transition to CONNECTING
2015-11-25 10:01:16,280 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 10:01:16,282 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 10:01:16,590 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 10:01:16,592 - kazoo.client - WARNING - Connection dropped: socket connection broken
2015-11-25 10:01:16,978 - ./consume-many.py - ERROR - **********||**********
2015-11-25 10:01:16,978 - ./consume-many.py - ERROR - repr: ZookeeperConnectionLost()
2015-11-25 10:02:25,606 - pykafka.balancedconsumer - WARNING - Internal partition registry doesn't match ZooKeeper!
2015-11-25 10:14:45,738 - kazoo.client - WARNING - Connection dropped: outstanding heartbeat ping not received
2015-11-25 10:14:45,738 - kazoo.client - WARNING - Transition to CONNECTING
2015-11-25 10:16:26,401 - pykafka.balancedconsumer - WARNING - Internal partition registry doesn't match ZooKeeper!
```

**Comment by yungchin** [comment]
@tigran-a thanks for providing further output, that's really helpful.

The last error you posted, the `ZookeeperConnectionLost`, is due to a that change we backed out in #351, so if you upgrade to pykafka 2.0.4 you should no longer encounter it.

The other error in [your earlier comment][0] however is something that we probably still need to fix. (And the first thing we need to do is fiddle with our logging so that we can see what type of exception was raised.)

[0]: https://github.com/Parsely/pykafka/issues/369#issuecomment-159530845

As for https://github.com/python-zk/kazoo/issues/357 - I believe the `KazooClient` will keep trying to re-establish the connection, so I suspect all we'd need to do is make sure that the consumer won't crash on `ConnectionLoss` exceptions, and that instead it will wait patiently until the connection comes back.

**Comment by yungchin** [comment]
(For later reference: [as per their docs][0], kazoo considers `ConnectionLoss` et al to be part of their private API, so we might have to check what exactly we should catch - and also, which of their errors are considered intermittent and which are permanent, so we don't swallow the latter too)

[0]: http://kazoo.readthedocs.org/en/2.2.1/api/exceptions.html

**Comment by tigran-a** [comment]
Just add some more errors that I'm getting:
```
2015-11-26 14:29:26,037 - pykafka.cluster - ERROR - Socket disconnected during offset manager discovery. This can happen when using PyKafka with a Kafka version lower than 0.8.2.
```
while Kafka is 0.8.2.2

Also got twice more this one: 
```
2015-11-27 01:35:40,362 - pykafka.balancedconsumer - ERROR - Exception encountered in worker thread:
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 238, in checker
    if not self._check_held_partitions():
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 607, in _check_held_partitions
    zk_partitions = self._get_held_partitions()
  File "/usr/local/lib/python2.7/dist-packages/pykafka/balancedconsumer.py", line 588, in _get_held_partitions
    all_partitions = self._zookeeper.get_children(self._topic_path)
  File "/usr/local/lib/python2.7/dist-packages/kazoo/client.py", line 1079, in get_children
    return self.get_children_async(path, watch, include_data).get()
  File "/usr/local/lib/python2.7/dist-packages/kazoo/handlers/utils.py", line 72, in get
    raise self._exception
```
In one case followed by
SessionExpiredError()

**Comment by yungchin** [comment]
Thanks @tigran-a - the offset manager discovery error is probably harmless, because as of #327 it will attempt reconnecting. We should probably downgrade it to a warning, I'll take care of that.

It seems `SessionExpiredError` is yet another "private API" exception out of kazoo, a more severe progression of `ConnectionLoss` I suppose.

Do you see these a lot? I'm not sure what the cause of this is, but certainly you're not alone as @amontalenti reported seeing a lot of lost zk sessions recently. Can I ask how many pykafka clients vs kafka brokers vs zk nodes you run?

**Comment by tigran-a** [comment]
SessionExpiredError took all my hard drive...  i just had a loop.. like `try consummer.consume() except print error` so it printed it in a loop.

my setup was: 
 - 2 running kafka brokers, 
 - topic with 2 partitions, replication factor 2
 - 2 clustered zookeeper hosts
 - two pykafka consumers set up to use those two zookeepers and 3 kafka brokers (one is unavailable, config left from a previous test)
everything was running in AWS

**Comment by yungchin** [comment]
@tigran-a thanks, hmmm just two consumers shouldn't be a problem for zookeeper - so I really wonder why the session seems to get suspended so often. But anyway, we should handle the pykafka end of this, and not break on session suspensions.

> SessionExpiredError took all my hard drive
It just occurred to me that this may not be very clear from our docs currently: if `consume()` raises an exception like that (which is an exception that's re-raised from a defunct worker thread), it will not recover from that. That is, the consumer instance is practically dead, and if you call `consume()` again you'll be handed the same exception again. 

(We could probably make this more apparent, maybe by wrapping re-raised exceptions in some custom exception derived from `ConsumerStoppedException`, something like that.)

To recap the bug that causes the exception: in `_check_held_partitions()` we currently don't handle the situation where our zookeeper session is temporarily suspended, causing the consumer to abort completely. 

( @emmett9001 I still hate the concept of  `_check_held_partitions()` - it sort of patches over errors we might have elsewhere, and I wonder if we still need it at all, now that we hardly cache any zk state anymore. So my proposed fix here would be to throw out the checker altogether :))

**Comment by emmett9001** [comment]
@yungchin I'm also considering whether `_check_held_partitions` is a necessary safeguard. All it's checking is that the internal consumer holds the same offsets as zookeeper, which is an invariant that should be maintained by the proper operation of the zookeeper `ChildWatch` callbacks. If those fail, either by triggering late or not at all, we have no way of noticing that situation. `fetch()` calls for the wrong partitions will continue uninterrupted, which is a big problem. I can see how `_check_held_partitions` masks other errors, but I'm not totally convinced that it's safe to remove it.

**Comment by yungchin** [comment]
@emmett9001 if the `ChildrenWatch` callbacks fail though, `_check_held_partitions` won't be able to detect that? Say you add a new `BalancedConsumer` to a group. An existing `BalancedConsumer` should now give up one or more of its partitions, but say for some reason `kazoo` doesn't run the callbacks. That would mean the existing consumer keeps ownership of all its zk nodes, and the new consumer will fail to register them, repeatedly. Then if the existing consumer runs `_check_held_partitions`, all that that sees is that all the partitions registered on zk are all the partitions held by the internal consumer, so it wouldn't correct the situation?

**Comment by emmett9001** [comment]
@yungchin Yes, these are good points. Any situation I can think of that involves the watches not being triggered will not cause a rebalance through any other means. Since rebalances are the only way that zookeeper partition sets are updated, watches not being triggered does not cause ZK partition sets to change. Since I can't think of any other way for the internal partition set to fall out of sync, I suppose I'll try testing a PR that removes the checker thread and see what comes of it.

[This](https://github.com/Parsely/pykafka/issues/163) is the issue that originally added the checker thread. @kbourgoin's main concern was exactly what we've been discussing here, and I believe we've shown that a missed `ChildrenWatch` can't actually cause the held set to de-sync.
